src/build/specialize_types/IR.zig

```diff
@@ -125,129 +147,157 @@ pub const Expr = union(enum) {
+    /// Reference to non-empty sublist into WhenBranch.List
     pub const NonEmptySlice = List.NonEmptySlice;
 };
 
-/// todo
+/// e.g. `|x| x * 2`
 pub const Function = struct {
     args: Pattern.Slice,
     return_type: Type.Idx,
+    /// ??? Is expr the body of the function?
     expr: Expr.Idx,
 };
 
-/// todo
+/// ??? For Record destructuring, e.g. `{ x, y } = coords(point)`
 pub const StructDestruct = struct {
+    /// ???
     ident: Ident.Idx,
+    /// ???
     field: Ident.Idx,
     kind: Kind,
-    /// todo
+    /// ???
     pub const Kind = union(enum) {
+        /// ???
         required,
+        /// ???
         guard: Pattern.Typed,
     };
-    /// todo
+    /// List of StructDestruct
     pub const List = collections.SafeMultiList(@This());
-    /// todo
+    /// Reference to sublist into StructDestruct.List
     pub const Slice = List.Slice;
 };
 
-/// todo
+/// Represents a pattern used in pattern matching e.g. `Ok x` as part of a when branch.
 pub const Pattern = union(enum) {
+    /// e.g. `x`
     identifier: Ident.Idx,
+    /// e.g. `.. as tail` in `[head, .. as tail] -> prepend(head, tail)`
     as: struct {
+        /// e.g. `..` in `.. as tail`
         pattern: Pattern.Idx,
+        /// e.g. `tail` in `.. as tail`
         ident: Ident.Idx,
     },
     str_literal: StringLiteral.Idx,
     number_literal: base.Literal.Num,
+    /// e.g. `Ok(x)` in `Ok(x) ->`
     applied_tag: struct {
+        /// Type of tag union e.g. `Result`
         tag_union_type: Type.Idx,
         tag_name: Ident.Idx,
+        /// e.g. `x` in `Ok(x)`
         args: Pattern.Slice,
     },
+    /// e.g. `{x, y}` in `{x, y} -> Point(x, y)`
     struct_destructure: struct {
         struct_type: Type.Idx,
+        /// ??? e.g. `x` and `y` in `{x, y}`
         destructs: StructDestruct.Slice,
+        /// e.g. `..` in `{x, y, ..}`
         opt_spread: ?Pattern.Typed,
     },
+    /// e.g. [Foo, Bar, ..]
     list: struct {
         elem_type: Type.Idx,
+        /// e.g. Foo and Bar in `[Foo, Bar, ..]`
         patterns: Pattern.Slice,
-
+        /// refers to e.g. `.. as tail` in `[head, .. as tail]`
         opt_rest: ?struct {
+            /// position in list of `..`; e.g. 0 in `[.., Foo, Bar]`
             offset: u16,
+            /// e.g. tail in `.. as tail`
             name: ?Ident.Idx,
         },
     },
+    /// _ is used as a catch-all, it matches anything
     underscore,
     compiler_bug: Problem.Compiler,
-    /// todo
+    /// List of Pattern
     pub const List = collections.SafeList(@This());
-    /// todo
+    /// Index into Pattern.List
     pub const Idx = List.Idx;
-    /// todo
+    /// Sublist reference of Pattern.List
     pub const Slice = List.Slice;
-    /// todo
+    /// Sublist reference into Pattern.List that is guaranteed to be non-empty
     pub const NonEmptySlice = List.NonEmptySlice;
-    /// todo
+    /// ??? Type annotated pattern e.g. `x: U64`, do we properly support this in a when branch in Roc?
```

Comment:
Once type solving has happened (which we're way past), everything will know its type. You might be right that it's not needed for the output of type specialization, though.